Adds Krav Maga

PKPenguin321 opened the pull request:
Code salvaged from https://github.com/tgstation/-tg-station/pull/10306

**Big update, this whole thing has been rewritten**
- Krav Maga is a fighting style, sort of like Sleeping Carp. Instead of using intent-based combos, however, it gives you a list of action buttons. Pressing one will cause your next attack to be whichever attack you selected.
- Krav Maga is currently admin-only. It can be used via the Krav Maga Gloves, which remove your Krav Maga if you remove them, and the Krav Maga implant, a single use implant that imbues the user with permanent Krav Maga.

_Krav Maga has three moves:_
1. Leg Sweep: A formidable stun. It sweeps the victims legs out from under them. Only usable on standing opponents.
2. Neck Chop: A devastating blow to the neck that prevents the victim from speaking for a while.
3. Lung Punch: A quick, precise, and powerful blow above the abdomen renders the victim unable to breathe for a short while.

_Other things that Krav Maga does:_
- Krav Maga makes your punches do a set 10 damage to opponents, and 15 if they're prone. However, you cannot get weaken punches.
- Krav Maga makes disarmed items go directly into your hands instead of onto the floor. However, you cannot push.
- Attacking with Krav Maga has unique sound effects to alert passerby of your fighting style.
- Preparing to use a move will alert passerby with a message, such as "Fighter McKrav assumes the Neck Chop stance!"

![Preview images](http://puu.sh/ojF9C/456ab649b3.png)
Sprites for the action buttons are by yours truly

PKPenguin321 committed:
adds krav maga

Zilenan91 commented:
This getting added to roundstart security? I know that security on Hippie has this to deal with the tide.

PKPenguin321 commented:
Implementations can be discussed later, this PR is purely for getting the code into the game

Fox-McCloud commented on code/datums/martial/krav_maga.dm:
```diff
@@ -0,0 +1,211 @@
+	H.verbs += /mob/living/carbon/human/proc/neck_chop
+	H.verbs += /mob/living/carbon/human/proc/head_elbow
+	H.verbs += /mob/living/carbon/human/proc/leg_sweep
+	H.verbs += /mob/living/carbon/human/proc/quick_choke
```
these all need to be spells or something---I dunno. Verbs are gross; action buttons are better.

PKPenguin321 replied:
I don't know, I think this system works really well. It was much more pleasant to use than I expected when I was testing it.

RemieRichards replied:
Nope, no verbs, Action buttons please.

PKPenguin321 replied:
RIP

Iamgoofball replied:
I'll set these up as an action button.

lzimann replied:
Even more action buttons... why

Iamgoofball replied:
we're not going to live in the dark ages because you live in it

lzimann replied:
sometimes i'm just annoyed that there is a fuckton of action buttons, like, the entire line full(especially when you use the cham kit)

drynwyn replied:
You can click and drag them to move them around, which I find helps.

RemieRichards commented on code/datums/martial/krav_maga.dm:
```diff
@@ -0,0 +1,211 @@
+	H.verbs -= /mob/living/carbon/human/proc/leg_sweep
+	H.verbs -= /mob/living/carbon/human/proc/quick_choke
+/datum/martial_art/krav_maga/proc/check_streak(var/mob/living/carbon/human/A, var/mob/living/carbon/human/D)
+	if(streak != "")
```
this is unneeded.

PKPenguin321 replied:
just this line or this entire proc? as far as i can tell the whole proc can be removed since i don't think this actually uses streaks

RemieRichards replied:
>as far as I can tell the whole proc can be removed since I don't think this actually uses streaks

tell that to the 4 verbs that all set streak to a value that is then used here to decide which attack is done

RemieRichards commented on code/datums/martial/krav_maga.dm:
```diff
@@ -0,0 +1,211 @@
+	if(slot == slot_gloves)
+		var/mob/living/carbon/human/H = user
+		style.teach(H,1)
+	return
```
trailing return, remove it please.

PKPenguin321 replied:
will do
this is an artifact of almost-year-old-goofcode so :ok_hand:

RemieRichards commented on code/datums/martial/krav_maga.dm:
```diff
@@ -0,0 +1,211 @@
+	var/mob/living/carbon/human/H = user
+	if(H.get_item_by_slot(slot_gloves) == src)
+		style.remove(H)
+	return
```
trailing return, remove it please.

Iamgoofball replied:
No

PKPenguin321 replied:
ill do it dont worry

xxalpha commented:
We need less martial arts, not more. The system is clunky to use and wasn't made to be a proper feature.

Iamgoofball commented:
@xxalpha if you actually read the PR you'd see that this isnt using the same system sleeping carp/plasma fist use

PKPenguin321 commented:
>balance tag

>on an admin only feature

PKPenguin321 committed:
addresses minor feedback, fixes implant
krav maga implant now works 100%
trailing returns and a useless if() removed

PKPenguin321 commented on code/datums/martial.dm:
```diff
@@ -7,6 +7,7 @@
 	var/datum/martial_art/base = null // The permanent style
 	var/deflection_chance = 0 //Chance to deflect projectiles
 	var/help_verb = null
+	var/counter_prob = 0//probability to counter direct melee attacks
```
I'm thinking I'm gonna just tear out all the countering shit, since it doesn't work, goof has given up on working on it, and it's probably unbalanced anyways. As far as I can tell, I can remove all the changes to this file.

Iamgoofball replied:
Oh, I can fix counters, I just dont have time to fix action buttons

PKPenguin321 replied:
sorry goofball you are too late

Iamgoofball replied:
i'll add it in post because it was cool

PKPenguin321 commented on code/datums/martial/krav_maga.dm:
```diff
@@ -0,0 +1,207 @@
+/datum/martial_art/krav_maga
+	name = "Krav Maga"
+	counter_prob = 25
```
Remove this for removing counters

PKPenguin321 commented on code/datums/martial/krav_maga.dm:
```diff
@@ -0,0 +1,207 @@
+						 	"<span class='userdanger'>You block the hit!</span>")
+			playsound(get_turf(A), 'sound/effects/hit_block.ogg', 50, 1, -1)
+			return 1
+	return 0
```
Remove all this up to line 5

PKPenguin321 commented on code/datums/martial/krav_maga.dm:
```diff
@@ -0,0 +1,207 @@
+		var/datum/martial_art/MA = D.martial_art
+		if(MA.on_hit(D,A)) // they countered with something
+			add_logs(A, D, "countered or blocked")
+			return 1
```
I think I can remove this up to line 106

PKPenguin321 commented on code/datums/martial/krav_maga.dm:
```diff
@@ -0,0 +1,207 @@
+	D.visible_message("<span class='danger'>[A] [picked_hit_type] [D]!</span>", \
+					  "<span class='userdanger'>[A] hits you!</span>")
+	var/bonus_damage = 10
+	if(D.weakened)
```
Should make this check for all prone states, not just weakened

PKPenguin321 commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -875,7 +875,9 @@
-
+				var/datum/martial_art/MA = H.martial_art
+				if(MA.on_hit(H,M)) // they countered with something
+					return
```
Can remove changes to this file

PKPenguin321 commented:
I should also be able to remove the block sound file.

Other things to do:

1. Rework verbs into action buttons (I'm not certain how at the moment, but you can be sure I'll study and test existing code until I get it)

2. Remove head elbow

3. Rename quick choke to lung punch, adjust description and related messages to make the losebreath clearer

4. Add actual cooldowns for the moves maybe

5. Add a message when the user is preparing to use a move, adjust the disarm message so it's clear that the user snatches the item, add more attack verbs

6. Not sure if it's possible, but maybe if the user is attacking a prone opponent and getting the bonus damage, change attack verb to "stomped" for more user feedback

PKPenguin321 committed:
big update
removes all counter/blocking related code and the related sound effect
makes krav maga use action buttons instead of verbs
renamed quick choke to lung punch
reduced stun time on leg sweep
removed head elbow, removed the help button (obsoleted by tooltips)

PKPenguin321 commented:
Ready for review/merge

PKPenguin321 commented:
Guess I might as well tag some maintainers
@Cheridan @KorPhaeron @duncathan @RemieRichards @ChangelingRain @Jordie0608

PKPenguin321 committed:
prevents krav maga from being used while incapacitated

ChangelingRain commented on code/datums/martial/krav_maga.dm:
```diff
@@ -0,0 +1,171 @@
+	if(owner.incapacitated())
+		owner << "<span class='warning'>You can't use Krav Maga while you're incapacitated.</span>"
+		return
+	owner << "<b><i>Your next attack will be a Leg Sweep.</i></b>"
```
These can be part of the below visible message(s) by having them as a second argument, ie `owner.visible_message("<span class='danger'>[owner] assumes the Leg Sweep stance!</span>", "<b><i>Your next attack will be a Leg Sweep.</i></b>")`

ChangelingRain commented on code/datums/martial/krav_maga.dm:
```diff
@@ -0,0 +1,171 @@
+
+/datum/martial_art/krav_maga/proc/neck_chop(var/mob/living/carbon/human/A, var/mob/living/carbon/human/D)
+	D.visible_message("<span class='warning'>[A] karate chops [D]'s neck!</span>", \
+				  	"<span class='userdanger'>[A] karate chops your neck, rendering you unable to speak for a short time!</span>")
```
'for a short time' really breaks the flow of this message and it'll sound pretty weird

PKPenguin321 commented on code/datums/martial/krav_maga.dm:
```diff
@@ -0,0 +1,171 @@
+/datum/martial_art/krav_maga//experimental
```
whoops, this //experimental is accidentally copied from my testing build

PKPenguin321 replied:
you're shitting me i forgot to remove it AGAIN

PKPenguin321 committed:
final style changes

PKPenguin321 committed:
removes experimental tag
